#7251 Vignette fread fwrite FR translated (#7253)
Rdatatable/data.table: vignettes/fr/datatable-fread-and-fwrite.Rmd

>
> — Matt Dowle  

Voir cet [exemple](https://stackoverflow.com/questions/36256706/fread-together-with-grepl/36270543#36270543) pour d'autres détails.

Sous Windows, les outils en mode ligne de commande tels que `grep` sont disponibles dans divers environnements, tels que Rtools, Cygwin, ou Windows Subsystem for Linux (WSL). Sous Linux et macOS, ces outils sont typiquement inclus dans le système d'exploitation.

#### 1.1.1  Lecture directe à partir d'une chaîne textuelle

`fread()` peut lire les données directement à partir d'une chaîne de caractères dans R en utilisant l'argument `text`. Ceci est particulièrement pratique pour créer des exemples reproductibles, pour tester des lignes de code, ou pour travailler avec des données générées par un programme à l'intérieur d'une session R. Chaque ligne de la chaîne doit être séparée avec le caractère de passage à la ligne `\n`.

```{r}
my_data_string = "colA,colB,colC\n1,apple,TRUE\n2,banana,FALSE\n3,orange,TRUE"
dt_from_text = fread(text = my_data_string)
print(dt_from_text)
```

#### 1.1.2 Lecture à partir d'URLs

> `fread()` peut lire les données directement à partir d'URLs web en passant l'URL en tant que chaîne de caractères dans l'argument `file`. Cela vous permet de télécharger et de lire les données à partir d'internet en une seule passe.

Author: Here just to explain word 'passe' = from word  passage = one step = one shot . Usuel en informatique et selon le dictionnaire Larousse français : ->  Passage en machine au cours duquel un traitement élémentaire, entrant dans UN processus complexe, est appliqué à un ensemble de données.  
ok, replaced by 'ligne'

  ```{r}
  # dt = fread("https://people.sc.fsu.edu/~jburkardt/data/csv/airtravel.csv")
  # print(dt)
  ```

#### 1.1.3 Décompression automatique des fichiers compressés

Dans beaucoup de cas, `fread()` peut automatiquement détecter et décompresser les fichiers directement en utilisant les extensions de compression communes et sans avoir besoin d'un objet de connexion explicite ou de commandes du shell. Il se base sur l'extension du fichier.

**Les extensions reconnues incluent typiquement :**
- `.gz` / `.bz2` (gzip / bzip2) : acceptés et fonctionnent de manière indépendante.
- `.zip` / `.tar`  (archives ZIP ou tar, fichier unique) : acceptées — `fread()` lira le premier fichier de l'archive si elle n'encontient qu'un seul.

> > Note : si l'archive contient plusieurs fichiers, `fread()` échouera avec une erreur.

Author: As it was the EN version. Changed here in FR into    **Note** :   to be coherent with similar previous  remarks in other vignettes, and reported in EN as   **Note**: 
ok, done.

Si on peut interpréter tous les champs de cette ligne comme des caractères et que les valeurs ne ressemblent pas trop à une ligne de données (par exemple une ligne de valeurs purement numériques pourrait encore être considérée comme des données), alors il s'agit typiquement d'une ligne d'entête (le nom des colonnes).

Sinon (c'est à dire si la ligne contient des types numériques reconnus, ou des chaînes de caractères qui ressemblent fortement à des nombres et qui pourraient être des données), elle est traitée comme une ligne de données et reçoit les noms de colonnnes par défaut (`V1`, `V2`, …).

Vous pouvez dire explicitement à fread si une entête existe en utilisant `header = TRUE`, ou pas `header = FALSE`.

**Détection des sauts de lignes**

Par défaut (`skip="auto"`), `fread` va automatiquement sauter les lignes vides ainsi que les lignes de commentaires (celles qui commencent par `#`) avant l'entête des données. Pour forcer manuellement un nombre donné de lignes, utilisez use `skip=n`.

### 1.3 Détection avancée et automatique du type de colonne

> Dans le monde réel, beaucoup d'ensembles de données contiennent des colonnes qui sont vides au départ, remplies avec des zéros, ou qui apparaissent numériques mais qui contiendront des caractères ultérieurement. Pour gérer de telles incohérences, `fread()` de `data.table` utilise une stratégie robuste de détection du type de colonne.

Author: - terms substitued
- in data.table removed because of the vignette context. Reported back in the EN version.

done.

Depuis la v1.10.5, `fread()` échantillonne les lignes en lisant des blocs de lignes contigües à partir de plusieurs points espacés régulièrement dans l'ensemble du fichier, y compris le début, le milieu et la fin. Le nombre de lignes échantillonnées est choisi dynamiquement en fonction de la taille et de la structure du fichier et vaut typiquement aux environs de 10 000, mais il peut être plus petit ou légèrement supérieur. Ce grand échantillonnage aide à détecter les changements de type qui se produisent ultérieurement dans les données (par exemple `001` qui devient `0A0`, ou des blancs qui deviennent des valeurs).

**Accès performant aux fichiers avec mmap**

> Pour implémenter efficacement cet échantillonnage, `fread()` utilise le mappage en mémoire de l'accès aux fichiers du système d'exploitation (`mmap`), ce qui permet de sauter à des positions arbitraires dans le fichier sans le parcourir séquentiellement. Cette stratégie paresseuse à la demande rend l'échantillonnnage presque instantané même avec des fichiers très gros.

Reviewer: paresseuse à la demande ?

> Si la destination du saut se trouve dans un champ entre guillemets (quotes) comprenant des passages à la ligne, `fread()` teste les lignes qui suivent jusqu'à trouver 5 lignes consécutives avec le nombre attendu de champs, pour permettre une analyse correcte même avec des fichiers complexes.

Reviewer: est-ce que ça marche avec les guillemets français ?

Author: The problem is that french language does not have the notion of _quote_. The FR keyboard uses the apostrophe which is inclined on the right (as in l**'**amitié) but displays a vertical quote in this text (but not in Word for example). The guillemets of FR keyboard **"**  produce the double quotes on one character as in english.  So the translation problem for me is to identify the different  cases: simple quotes, double quotes, apostrophe, guillemets. In FR language we use also the EN terms so for anybody here  'entre simple quotes' and 'entre double quotes' is culturally understandable is we accept to use  _quote_ in the FR translation.  \" and \' remain usable as it concerns the protection of the same resulting characters .     

Question: so does _quote_ remain in the FR text instead of guillemets ?

Author: > est-ce que ça marche avec les guillemets français ?

il faudrait développer le paragraphe à l'aide d'un exemple (peut être demander à l'auteur ce qu'il voulait dire).

**Détection précise et optimisée du type**

Le type de chaque colonne est déduit en se basant sur le type le plus bas nécessaire de la liste ordonnée suivante :

`booléen` < `entier` < `entier 64` < `double` < `caractère`

Ce qui permet :

- d'allouer la mémoire de manière unique à l'avance en utilisant le type correct

> - d'éviter la relecture du fichier ou de définir les colClasses manuellement 

Author: ok, backticks added

done.

- d'améliorer la vitesse et l'efficacité de la mémoire

> **Exceptions sur le type dans les valeurs non échantillonnées**

Reviewer: ```suggestion
**Différents types dans les valeurs non échantillonnées**
```

Si le type est modifié dans les lignes qui ne sont pas échantillonnées, `fread()` détecte cela automatiquement et relit le fichier pour rétablir l'assignation correcte du type, sans nécessiter l'intervention de l'utilisateur. Par exemple une colonne échantillonnée comme entier pourrait ultérieurement contenir `00A` — ce qui déclenchera une relecture automatique en tant que caractères.

En activant `verbose=TRUE` vous pourrez voir toute la logique de détection ainsi que les relectures.

> ### 1.4 Détection des erreurs au plus tôt à la fin du fichier

Reviewer: ```suggestion
### 1.4 Détection des erreurs à la fin du fichier au plus tôt 
```

Author: ok swapped.

done.

> Parce qu'un grand nombre d'échantillons signifie que l'on se rapproche de la fin du fichier, les problèmes critiques tels qu'un nombre incohérent de colonnes, un bas de page mal formaté, ou des guillemets ouvrants qui ne sont pas fermés — peuvent être détectés et rapportés presque instantanément. Cette détection d'erreurs au plus tôt évite la surcharge d'un nouveau traitement du fichier complet ou l'allocation excessive de la mémoire, simplement pour trouver une erreur à la fin. Cela permet une réaction plus rapide et une utilisation plus efficace des ressources, particulièrement lorsque vous travaillez avec de gros ensembles de données.

Reviewer: ```suggestion
Parce que l'échantillon contient la fin du fichier, les problèmes critiques tels qu'un nombre incohérent de colonnes, un bas de page mal formaté, ou des guillemets ouvrants qui ne sont pas fermés — peuvent être détectés et rapportés presque instantanément.
Cette détection d'erreurs au plus tôt évite la surcharge d'un nouveau traitement du fichier complet ou l'allocation excessive de la mémoire, simplement pour trouver une erreur à la fin.
Cela permet une réaction plus rapide et une utilisation plus efficace des ressources, particulièrement lorsque vous travaillez avec de gros ensembles de données.
```

Author: - text corrected
- setting new lines and reported to EN version

done.

options(datatable.integer64 = "double")   # Exemple : définir globalement à "double"
getOption("datatable.integer64") 
```

### 1.6 Ignorer ou sélectionner les colonnes par nom ou par position

Pour économiser de la mémoire et pour améliorer les performances, utilisez les arguments `select` ou `drop` de `fread()` pour lire uniquement les colonnes dont vous avez besoin.

- Si vous avez besoin uniquement de quelques colonnes, utilisez `select`.
- Si vous voulez exclure uniquement quelques colonnes, utilisez `drop` — ceci évite de lister tout ce que vous voulez garder.

Points clé:
- `select` : vecteur des noms des colonnes ou des positions à garder (les autres seront ignorés).
- `drop`: vecteur des noms des colonnes ou des positions à ignorer (les autres sont gardés).
- N'utilisez pas `select` et `drop` simultanément — ils sont mutuellement exclusifs.
- `fread()` vous avertira si une colonne que vous avez spécifiée n'existe pas dans le fichier.

Pour les détails, voir la page du manuel en exécutant `?fread` dans R.

### 1.7 Sauter à une ligne de l'entête d'une sous-table en utilisant une sous-chaîne d'un nom de colonne

> Utilisez `skip="chaîne"` dans `fread` pour chercher une ligne contenant une sous-chaîne (typiquement issue des noms de colonne, par exemple `skip="Date"`). La lecture commence à la première ligne où la correspondance est trouvée. Ceci est utile pour sauter les métadonnées ou pour sélectionner des sous-tables dans les fichiers contenant des tables multiples. Cette fonctionnalité est inspirée de la fonction `read.xls` du paquet gdata .

Reviewer: I did not know about this feature, but I had to read the english version to understand what it does.
can this title be clarified?

Reviewer: OK, in https://github.com/Rdatatable/data.table/pull/7253/commits/748631bb9cbf916c5daa1492227cd176edbd578e I deleted that section heading, and moved the corresponding text up to the other section which discusses the skip defaults. Does that make sense to you?

### 1.8 Détection automatique des guillemets (échappés ou pas)

> `fread` détecte automatiquement la manière dont les guillemets sont échappés — qu'ils soient simples ('') ou doubles ("") ou échappés avec la barre oblique inverse (respectivement \' et \") — sans nécessiter l'intervention de l'utilisateur. Ceci est déterminé en utilisant un grand échantillon de données (voir le point 3), et vérifié sur le fichier complet.

Reviewer: écrire que les guillemets français ne marche pas ?

Author: normalement ça marche (voir mon explication ci-dessus)  car nos caractères apostrophe et guillemets du clavier FR sont  directement affichés sous la forme anglaise comme lorsque l'on entre du code standard. 

Open point: je suis sur tablette actuellement mais j'essaierai avec R sur le desktop la semaine prochaine.

> Scénari pris en charge :

Reviewer: ```suggestion
Exemples pris en charge :
```

- Guillemets non échappés dans les champs entre guillemets

> par exemple, `"This "quote" is invalid"` — est reconnu tant que le nombre de colonnes reste cohérent.

Reviewer: ````suggestion
par exemple, `"Ces "guillemets" ne sont pas valides, mais fread marche quand même"` — est reconnu tant que le nombre de colonnes reste cohérent :
```{r}
data.table::fread(text='x,y\n"Ces "guillemets" ne sont pas valides, mais fread marche quand même",1')
```
```
````

Author: ok corrected.

done.

- Champs qui ne sont pas entre guillemets mais qui commencent par des guillemets

> par exemple, `Invalid"Field,10,20` — est reconnu correctement comme n'étant pas un champ entre guillemets.

Reviewer: ````suggestion
par exemple, `pas"valide,1` — est reconnu correctement comme n'étant pas un champ entre guillemets.
```{r}
data.table::fread(text='x,y\npas"valide,1')
```
````

Contraintes et limitations :
- les règles de l'échappement et du nombre des colonnes doivent être cohérentes tout au long du fichier.

- non reconnu lorsque `fill=TRUE` — dans ce cas le fichier doit suivre la RFC4180 concernant les guillemets et l'échappement.

Robustesse selon la version :
A partir de la v1.10.6, `fread` résoud les ambiguïtés de manière plus fiable au travers du fichier entier en utililisant la cohérence du nombre entier de colonnes (valeur par défaut `fill=FALSE`). Les avertissements sont émis si l'analyse échoue par ambiguité.

## 2. fwrite()

> `fwrite()` est le pendant de `fread()` pour l'enregistrement rapide. Il est conçu pour être rapide, sensible aux defaults, et facile à utiliser, tout en reprenant la plupart des facilités de `fread`.

Reviewer: ```suggestion
Pour l'écriture rapide d'un fichier CSV, `fwrite()` est le partenaire de `fread()`.
Il est conçu pour être rapide, sensible aux defaults, et facile à utiliser, tout en reprenant la plupart des facilités de `fread`.
```

```

### 2.2 Sérialisation à granularité fine pour la datet et l'heure (dateTimeAs)

Permet le contrôle précis pour POSIXct et les types de Date :

- `dateTimeAs="ISO"` (valeur par défaut pour POSIXct) : format ISO 8601 (c'est à dire AAAA-MM-JJTHH:MM:SS.ffffffZ), ce qui préserve la précision après la virgule pour le portage non ambigu.

- `dateTimeAs="epoch"` : POSIXct en secondes depuis epoch (numérique).

```{r}
dt_timestamps = data.table(
  ts = as.POSIXct("2023-10-26 14:35:45.123456", tz = "GMT"),
  dt = as.Date("2023-11-15")
)
temp_dt_iso = tempfile(fileext = ".csv")
fwrite(dt_timestamps, temp_dt_iso, dateTimeAs = "ISO")
cat(readLines(temp_dt_iso), sep = "\n")
unlink(temp_dt_iso)
```

### 2.3 Gestion de bit64::integer64

**Précision complète pour les grands entiers** : `fwrite` écrit les colonnes `bit64::integer64` en les convertissant en chaînes de caractères avec la précision complète. Ce qui évite la perte de données ou la conversion silencieuse en double qui pourrait se produire avec des écrivains moins spécialisés. Ceci est crucial pour les IDs ou les mesures qui demandent plus que l'intervalle des entiers `32-bit` du standard R ou de la double précision `53-bit`.

> **Gestion directeg** : ce traitement direct et soigné des nombres spéciaux assure l'intégrité des données et l'efficacité des entrées / sorties, en évitant les conversions intermédiaires inutiles vers des types moins précis.

Author: plutôt  Gestion directe  à cause du feminin.

done.

```{r}
if (requireNamespace("bit64", quietly = TRUE)) {
  dt_i64 = data.table(uid = bit64::as.integer64("1234567890123456789"), val = 100)
  temp_i64_out = tempfile(fileext = ".csv")
  fwrite(dt_i64, temp_i64_out)
  cat(readLines(temp_i64_out), sep = "\n")

> 

Reviewer: ```suggestion
```

Author: ok line  deleted

done.

  unlink(temp_i64_out)
}
```

### 2.4 Ordre des colonnes et contrôle des sous-ensembles

> Pour contrôler l'ordre et le sous-ensemble des colonnes écrites dans le fichier, réduisez le data.table avant d'appeler `fwrite()`. L'argument `col.names` de `fwrite()` est un booléen (TRUE/FALSE) qui contrôle si la ligne d'entête est écrite, et non quelles colonnes sont écrites.

Reviewer: ```suggestion
Pour contrôler l'ordre et le sous-ensemble des colonnes écrites dans le fichier, convertir le `data.table` avant d'appeler `fwrite()`. L'argument `col.names` de `fwrite()` est un booléen (TRUE/FALSE) qui contrôle si la ligne d'entête est écrite, et non quelles colonnes sont écrites.
```

Author: corrected and backticks added

done.

```{r}
dt = data.table(A = 1:3, B = 4:6, C = 7:9)

# écrire uniquement les colonnes C et A, dans cet ordre
fwrite(dt[, .(C, A)], "out.csv")
cat(readLines("out.csv"), sep = "\n")
file.remove("out.csv")
```

## 3. Note sur les performances

> Bien que cette vignette se concentre sur les fonctionnalités et sur l'utilisation, le but de `fread` et `fwrite` est la vitesse. Les performances des E/S de `data.table` sont un sujet pour le banc d'essais continu.

Reviewer: ```suggestion
Bien que cette vignette se concentre sur les fonctionnalités et sur l'utilisation, le but de `fread` et `fwrite` est la vitesse.
```

> Pour les utilisateurs intéressés par les détails, et les performances comparées mises à jour, nous recommandons ces billets de blog externes, utilisant le paquet `atime` pour une analyse rigoureuse :

Author: ok corrected.

done.

- **[data.table asymptotic timings](https://tdhock.github.io/blog/2023/dt-atime-figures/)** : compare les performances de `fread` et `fwrite` par rapport à d'autres paquets populaires de R tels que `readr` et `arrow`.

> - **[Benchmarking data.table with polars, duckdb, and pandas](https://tdhock.github.io/blog/2024/pandas-dt/)** : compare les E/S de `data.table` et regroupe les performances face aux bibliothèques principales Python.

Reviewer: ```suggestion
- **[Benchmarking data.table with polars, duckdb, and pandas](https://tdhock.github.io/blog/2024/pandas-dt/)** : compare avec bibliothèques principales Python.
```

> Ces bancs d'essais montrent d'une manière cohérente que `fread` et `fwrite` sont hautement compétitifs et occupent souvent les premières places quant aux performances dans l'écosystème R.

Author: ok corrected.

done.
